core/modules/loader/MasterServer.cc

```diff
@@ -0,0 +1,277 @@
+                                               boost::asio::ip::udp::endpoint const& senderEndpoint) {
+    std::string funcName("MasterServer::workerListRequest");
+    LOGS(_log, LOG_LVL_DEBUG, funcName);
+
+    int tcpPort = 0; // only needed for create parameter
+    auto addr = NetworkAddress::create(data, tcpPort, funcName);
+    if (addr == nullptr) {
+        std::string errStr("STATUS_PARSE_ERR parse error in " + funcName);
+        LOGS(_log, LOG_LVL_ERROR, errStr);
+        return replyMsgReceived(senderEndpoint, inMsg, LoaderMsg::STATUS_PARSE_ERR, errStr);
+    }
+
+    // TODO: put this in a separate thread.
+    auto workerList = _centralMaster->getWorkerList();
+    workerList->sendListTo(inMsg.msgId->element, addr->ip, addr->port, getOurHostName(), getOurPort());
+    LOGS(_log, LOG_LVL_INFO, funcName << " done sendListTo ");
+
+    return nullptr;
+}
+
+
+BufferUdp::Ptr MasterServer::workerKeysInfo(LoaderMsg const& inMsg, BufferUdp::Ptr const& data,
+                                            boost::asio::ip::udp::endpoint const& senderEndpoint) {
+
+    std::string funcName("MasterServer::workerKeysInfo");
+    LOGS(_log, LOG_LVL_DEBUG, funcName);
+
+    try {
+        uint32_t name;
+        NeighborsInfo nInfo;
+        StringRange strRange;
+        ProtoHelper::workerKeysInfoExtractor(*data, name, nInfo, strRange);
+        LOGS(_log, LOG_LVL_INFO, funcName << "name=" << name << " keyCount=" << nInfo.keyCount <<
+                                 " recentAdds=" << nInfo.recentAdds << " range=" << strRange);
+        // TODO move to separate thread.
+        _centralMaster->updateWorkerInfo(name, nInfo, strRange);
+    } catch (LoaderMsgErr &msgErr) {
+        LOGS(_log, LOG_LVL_ERROR, msgErr.what());
+        return replyMsgReceived(senderEndpoint, inMsg, LoaderMsg::STATUS_PARSE_ERR, msgErr.what());
+    }
+    return nullptr;
+}
+
+
+BufferUdp::Ptr MasterServer::workerInfoRequest(LoaderMsg const& inMsg, BufferUdp::Ptr const& data,
+                                 boost::asio::ip::udp::endpoint const& senderEndpoint) {
+    // TODO HIGH Wrap this up in a command and put it on a queue.
+    try {
+        std::string const funcName("MasterServer::workerInfoRequest");
+        int tcpPort = 0;
+        NetworkAddress::UPtr requestorAddr = NetworkAddress::create(data, tcpPort, funcName);
+        if (requestorAddr == nullptr) {
+            throw LoaderMsgErr(ERR_LOC, "requestorAddr nullptr");
+        }
+
+        auto protoItem = StringElement::protoParse<proto::WorkerListItem>(*data);
+        if (protoItem == nullptr) {
+            throw LoaderMsgErr(ERR_LOC, "protoItem nullptr");
+        }
+
+        auto workerId = protoItem->wid();
+        LOGS(_log, LOG_LVL_INFO, funcName << " Master got wId=" << workerId);
+
+        /// Find the worker name in the map.
+        auto workerItem = _centralMaster->getWorkerWithId(protoItem->wid());
```

Comment:
Why not to use `workerId` which you got 2 lines above?
```
auto workerItem = _centralMaster->getWorkerWithId(workerId);
```

Reply:
Alternatively, just don't cache `workerItem` and use `protoItem->wid()` everywhere, especially if this operation is cheap.